### PR TITLE
Clean up, improve and unify configuration and option classes handling of time and size properties

### PR DESCRIPTION
**Pull request by medb**
Changes in this PR:
- support size suffixes in values of size properties using [`Configuration#getLongBytes`](https://hadoop.apache.org/docs/r3.3.4/api/org/apache/hadoop/conf/Configuration.html#getLongBytes-java.lang.String-long-) method
- support time suffixes in values of time properties using [`Configuration#getTimeDuration`](https://hadoop.apache.org/docs/r3.3.4/api/org/apache/hadoop/conf/Configuration.html#getTimeDuration-java.lang.String-long-java.util.concurrent.TimeUnit-) method
- migrate time properties in option classes to `Duration` type
- adjust default values of the properties:
  ```
  fs.gs.outputstream.upload.chunk.size (64m -> 24m)
  fs.gs.http.connect-timeout (20s -> 5s)
  fs.gs.http.read-timeout (20s -> 5s)
  ```
- remove inline default value constants in option classes
- misc configuration/option-related code cleanup

### CODE REVIEW

File: gcs/CONFIGURATION.md [hunk 1]
```diff
@@ -22,7 +22,7 @@
     Whether to perform copy operation using Rewrite requests which allows copy
     files between different locations and storage classes.
 
-*   `fs.gs.rewrite.max.bytes.per.call` (default: `536870912`)
+*   `fs.gs.rewrite.max.chunk.per.call` (default: `512m`)
 
     Maximum number of bytes rewritten in a single rewrite request when
```
Comment (arunkumarchacko) [inline gcs/CONFIGURATION.md:27]: Do we have to change the description to match the same? i.e. in the description, we are mentioning bytes and the value can be in MB for e.g.

File: gcs/CONFIGURATION.md [hunk 2]
```diff
@@ -22,7 +22,7 @@
     Whether to perform copy operation using Rewrite requests which allows copy
     files between different locations and storage classes.
 
-*   `fs.gs.rewrite.max.bytes.per.call` (default: `536870912`)
+*   `fs.gs.rewrite.max.chunk.per.call` (default: `512m`)
```
Comment (arunkumarchacko) [inline gcs/CONFIGURATION.md:25]: do we want to change this to chunksize?

File: gcs/CONFIGURATION.md [hunk 3]
```diff
@@ -22,7 +22,7 @@
     Whether to perform copy operation using Rewrite requests which allows copy
     files between different locations and storage classes.
 
-*   `fs.gs.rewrite.max.bytes.per.call` (default: `536870912`)
```
Comment (arunkumarchacko) [inline gcs/CONFIGURATION.md:25]: A general question. What is the issue with what is already present? i.e why fixing something which is not broken.

File: gcs/CONFIGURATION.md [hunk 4]
```diff
@@ -447,10 +445,10 @@ default service account impersonation.
     modifications made outside this connector instance may not be immediately
     reflected.
 
-*   `fs.gs.performance.cache.max.entry.age.ms` (default: `5000`)
```
Comment (arunkumarchacko) [inline gcs/CONFIGURATION.md:450]: if someone do not provide the unit, it will be taken as ms? i.e. Will customer need to change their configs once they migration to 3.0?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchHelper.java [hunk 5]
```diff
@@ -71,15 +71,15 @@ public class BatchHelper {
...
       return new BatchHelper(requestInitializer, gcs, maxRequestsPerBatch, /* numThreads= */ 0);
     }
 
     BatchHelper newBatchHelper(
         HttpRequestInitializer requestInitializer,
         Storage gcs,
-        long maxRequestsPerBatch,
-        long totalRequests,
+        int maxRequestsPerBatch,
+        int totalRequests,
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchHelper.java:82]: why changing this?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateFileOptions.java [hunk 6]
```diff
@@ -44,9 +44,9 @@ public enum WriteMode {
   public static Builder builder() {
     return new AutoValue_CreateFileOptions.Builder()
         .setAttributes(ImmutableMap.of())
-        .setContentType(CreateObjectOptions.CONTENT_TYPE_DEFAULT)
+        .setContentType(CreateObjectOptions.DEFAULT_OVERWRITE.getContentType())
         .setEnsureNoDirectoryConflict(true)
-        .setMinSyncInterval(Duration.ofSeconds(10))
+        .setMinSyncInterval(Duration.ZERO)
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateFileOptions.java:49]: why changing the default of 10 seconds?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java [hunk 7]
```diff
@@ -27,16 +27,14 @@
...
-
   public static final CreateObjectOptions DEFAULT_NO_OVERWRITE = builder().build();
   public static final CreateObjectOptions DEFAULT_OVERWRITE =
       builder().setOverwriteExisting(true).build();
 
   public static Builder builder() {
     return new AutoValue_CreateObjectOptions.Builder()
         .setContentEncoding(null)
-        .setContentType(CONTENT_TYPE_DEFAULT)
+        .setContentType("application/octet-stream")
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java:37]: why is this change required?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageOptions.java [hunk 8]
```diff
@@ -34,67 +34,6 @@
...
-  public static final String STORAGE_SERVICE_PATH_DEFAULT = Storage.DEFAULT_SERVICE_PATH;
-
-  /** Default setting for enabling auto-repair of implicit directories. */
-  public static final boolean AUTO_REPAIR_IMPLICIT_DIRECTORIES_DEFAULT = true;
-
-  /**
-   * Default setting for the length of time to wait for empty objects to appear if we believe we are
-   * in a race with multiple workers.
-   */
-  public static final int MAX_WAIT_MILLIS_FOR_EMPTY_OBJECT_CREATION = 3_000;
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageOptions.java:66]: why changing this? I think there the existing structure is more readble.

File: gcs/CHANGES.md [hunk 9]
```diff
@@ -82,7 +78,45 @@
...
+
+    ```
+    fs.gs.http.connect-timeout
+    fs.gs.http.read-timeout
+    fs.gs.max.wait.for.empty.object.creation
+    fs.gs.outputstream.sync.min.interval
+    fs.gs.performance.cache.max.entry.age
+    ```
+
+1.  Change default values of properties:
```
Comment (arunkumarchacko) [inline gcs/CHANGES.md:113]: why are we changing these?

**Review by arunkumarchacko — CHANGES_REQUESTED** [review]

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchHelper.java (hunk 5 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/BatchHelper.java:82]: For consistency - all constructor parameters are `int` now, and we do not need them to be `long`.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateFileOptions.java (hunk 6 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateFileOptions.java:49]: For consistency between GCS connector and GCSIO default settings.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java (hunk 7 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/CreateObjectOptions.java:37]: There is a little utility in having default constants for each option class filed, if we already have default instance of the option class itself.

File: gcs/CHANGES.md (hunk 9 above)
Comment (medb) [reply gcs/CHANGES.md:113]: Our perf team did analysis and 5s timeout values are mote performant. High memory consumption during write is a long-standing issue too - reducing buffer to 24m to mitigate it, in cost of minor perf impact for large file writes.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageOptions.java (hunk 8 above)
Comment (medb) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageOptions.java:66]: They provide no value as we already have default instance of option class itself, clutter the code, and they are not consistent between option classes.

File: gcs/CONFIGURATION.md (hunk 4 above)
Comment (medb) [reply gcs/CONFIGURATION.md:450]: Yes, it will be treated as `ms` - there are unit test for this too.

File: gcs/CONFIGURATION.md (hunk 3 above)
Comment (medb) [reply gcs/CONFIGURATION.md:25]: It's quite hard to use this properties - we constantly need to convert gb, mb, etc to bytes. Also, this is a standard functionality for Hadoop properties, so it could be surprising for users that they can not use units in properties.

File: gcs/CONFIGURATION.md (hunk 1 above)
Comment (medb) [reply gcs/CONFIGURATION.md]: Done

File: gcs/CONFIGURATION.md (hunk 2 above)
Comment (medb) [reply gcs/CONFIGURATION.md]: Done

**Review by arunkumarchacko — APPROVED** [review]

**Review by cnauroth — CHANGES_REQUESTED** [review]
@medb , thanks for the patch! This mostly looks good to me. I see there are some merge conflicts that need to be resolved. Aside from that, I just entered one comment with a potential refactoring idea. Let me know what you think.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystem.java [hunk 10]
```diff
@@ -542,8 +542,10 @@ public FSDataOutputStream create(
                                   : CreateFileOptions.WriteMode.CREATE_NEW)
                           .setMinSyncInterval(
                               Duration.ofMillis(
-                                  GCS_OUTPUT_STREAM_SYNC_MIN_INTERVAL_MS.get(
-                                      getConf(), getConf()::getInt)))
+                                  GCS_OUTPUT_STREAM_SYNC_MIN_INTERVAL.get(
+                                      getConf(),
+                                      (name, defVal) ->
```
[the remark this answers is not in the document]
Comment (medb) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystem.java]: Great suggestion! Done.
